### PR TITLE
`constrained-generators`: `(++.)` and `singleton_` for lists

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

This is a simplified approach to that of #4658 with the goal of just doing what's necessary for `++` and `singleton`. Some things haven't been finished yet:

- [x] clean up the implementation of `propagateSpecFun`
- [x] implement the fold spec cases
- [x] examples and better tests than just soundness
- [x] ensure a higher success ratio on `prop_univSound $ TestableFn (appendFn @BaseFn @Int)`
- [x] add syntax-level functions
- [x] comments + better error messages

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -4483,13 +4488,54 @@ instance BaseUniverse fn => Functions (ListFn fn) fn where
...
     FoldMap f | NilCtx HOLE <- ctx -> typeSpec (ListSpec Nothing [] TrueSpec TrueSpec $ FoldSpec f spec)
+    SingletonList | NilCtx HOLE <- ctx -> case spec of
+      TrueSpec -> TrueSpec
+      MemberSpec xss -> MemberSpec [a | [a] <- xss]
+      TypeSpec (ListSpec _ m sz e f) cant
+        -- TODO: fix these error messages!
+        | length m > 1 -> ErrorSpec $ pure "singleton 1"
+        | not $ 1 `conformsToSpec` sz -> ErrorSpec $ pure "singleton 2"
+        | FoldSpec{} <- f -> ErrorSpec $ pure "singleton 3" -- TODO: this is just lazyness for now
+        | any (not . (`conformsToSpec` e)) m -> ErrorSpec $ pure "singleton 4"
```
Comment (UlfNorell) [inline libs/constrained-generators/src/Constrained/Base.hs]: Do we not maintain ``all (`conformsToSpec` elemSpec) must``?

Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs]: Not in the soundness tests...

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 2]
```diff
@@ -4483,13 +4488,54 @@ instance BaseUniverse fn => Functions (ListFn fn) fn where
             let args = appendList (mapList (\(Value a) -> Lit a) pre) (v' :> mapList (\(Value a) -> Lit a) suf)
              in Let (App (injectFn fn) args) (x :-> p)
     FoldMap f | NilCtx HOLE <- ctx -> typeSpec (ListSpec Nothing [] TrueSpec TrueSpec $ FoldSpec f spec)
+    SingletonList | NilCtx HOLE <- ctx -> case spec of
+      TrueSpec -> TrueSpec
+      MemberSpec xss -> MemberSpec [a | [a] <- xss]
+      TypeSpec (ListSpec _ m sz e f) cant
+        -- TODO: fix these error messages!
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Base.hs]: Yes, fix them. Also make some tests that cause the messages, and inspect them to see that they are clear, and they contain enough information to understand what went wrong.

**Review by TimSheard — APPROVED** [review]
We need better error messages, that explain what went wrong.
E.g. "some member of the must set, 5, does not meet the constraint that all list elements must meet:  even x" 

Hard to tell, but better formatting may make the code easier to understand.
Also code comments that explain the strategy would be helpful.
The strategy is to construct a Specification that "fixes" each part of the ListSpec, to adjust for the known part of the context that cannot be part of the new Spec.

**Comment by MaximilianAlgehed** [comment]
Nothing much to do about the formatting, but some better error messages and comments are on my mental todo list...

**Pull request merged**